ouroboros-consensus/docs/future-ledger-hd.md

```diff
@@ -0,0 +1,258 @@
+```haskell
+data RewardUpdate c = RewardUpdate
+  { deltaT :: !DeltaCoin
+  , deltaR :: !DeltaCoin
+  , rs :: !(Map (Credential 'Staking c) (Set (Reward c)))
+  , deltaF :: !DeltaCoin
+  , nonMyopic :: !(NonMyopic c)
+  }
+```
+
+where we will use the `delta*` fields to update the treasury, reserves and fee
+pot on the `NEWEPOCH` rule. We also use the `rs` field when calling
+`updateRewards` which happens also on `NEWEPOCH`.
+
+
+If Consensus can compute the `RewardUpdate` (possibly on a separate thread that
+traverses the map at its own pace?) then we can provide the `RewardUpdate` to
+the `NEWEPOCH` rule so that the pots can be updated.
+
+> **_PLAN:_** Consensus will compute the rewards outside of the Ledger and will
+> provide the parts of the `RewardUpdate` to the ledger. In particular, on the
+> epoch boundary, it will provide `deltaT`, `deltaR`, and `deltaF`. Ledger will
+> not compute the `RewardUpdate` thus the pulser becomes dead code.
+
+> Moreover, the ledger will provide a function `f :: Block -> Set
+> (StakeCredential c)` so that Consensus can supply Ledger with only the part of
+> the `UMap` (the unified map) that it requires. In particular, Consensus will
+> have to apply the `rs :: !(Map (Credential 'Staking c) (Set (Reward c)))` field
+> of the reward update to the unified map on the epoch boundary, prior to
+> computing this view for the ledger (to ensure that reward withdrawals are
+> correct within the given block).
+
+## Snapshots and Leader Schedule
+
+The `checkIsLeader` functions for Praos and TPraos makes use of the stake
+distribution by stake pool in the `LedgerView` (see the definitions in
+`ouroboros-consensus-protocol`). If the snapshots (and therefore the
+`ssStakeMarkPoolDistr` field) reside in the Consensus side, we can produce the
+relevant stake distributions when needed and don't involve the ledger. In any
+case this functionality is in between Ledger and Consensus so it makes sense to
+move it out of the ledger.
+
+> **_PLAN:_** Consensus will manage the snapshots to produce stake distribution
+> by pool that can be used by Consensus later to resolve queries about the
+> LeaderSchedule. Ledger will not know about the Snapshots. In particular, the
+> UTxO-HD report includes the concept of Snapshots of tables, which would be
+> used to manage and access these snapshots.
+
+Note that this implies creating a new package or component at the
+Consensus-Ledger boundary whose owner would probably be the Consensus team as
+its responsibilities would be related with computations required for the
+Consensus protocol (leader checks, and similar).
+
+## Open questions
+
+- Should the Ledger return the diffs? it actually internally compute diffs, but
+  they the diffs are applied to the values before returning. If we wanted to
+  return the diffs instead, there are many intermediate layers through which
+  they will have to be floated, but it should be doable.
```

Comment:
If this refers to what I think it does, it's covered in the second report and in the prototype. The solution is to paramaterise the ledger rules by the table type, and then to use a "partial tracking map" which fits the map interface and also tracks the diffs. This gives us the diffs we want without any "real" changes in the ledger rules (just abstracting over the map interface for the on-disk tables).